File: test/vectors/cross-emitter-matrix/sign.py
#!/usr/bin/env python3
# Cross-emitter matrix — Python signer.
# Reads a record (JSON) from stdin; writes {canonical, signature_hex} to stdout.
# Uses tuple-array canonical form matching ../verify.py.
# Signing key: hex-encoded 32-byte Ed25519 secret from SIGNING_KEY_HEX env var.

import hashlib
import json
import os
import sys
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
from cryptography.hazmat.primitives import serialization


FIELD_ORDER = [
    "id", "timestamp", "method", "toolName", "namespace", "upstream",
    "principal", "durationMs", "success", "errorCode", "previousHash",
]


def assert_well_formed(value):
    for i, ch in enumerate(value):
        if 0xD800 <= ord(ch) <= 0xDFFF:
            raise ValueError(f"unpaired surrogate at index {i}")


def canonicalize_value(value):
    if value is None:
        return None
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        assert_well_formed(value)
        return value
    if isinstance(value, int):
        if abs(value) > 2**53 - 1:
            raise ValueError(f"unsafe number {value}")
        return value
    if isinstance(value, float):
        raise ValueError(f"unsafe number {value}")
    if isinstance(value, list):
        return ["L", [canonicalize_value(v) for v in value]]
    if isinstance(value, dict):
        keys = sorted(value.keys(), key=lambda k: k.encode("utf-16-be"))
        for k in keys:
            assert_well_formed(k)
        return ["M", [[k, canonicalize_value(value[k])] for k in keys]]
    raise ValueError(f"unsupported type {type(value)}")


def canonicalize(record):
    ordered = [[k, canonicalize_value(record.get(k))] for k in FIELD_ORDER]
    for optional in ("decisionContextDigest", "extensionsDigest", "aiInvocation", "parties"):
        if record.get(optional) is not None:
            ordered.append([optional, canonicalize_value(record[optional])])
    return json.dumps(ordered, ensure_ascii=False, separators=(",", ":"))


def daemon_main():
    # Env pinning: only set fallbacks; do not override the driver.
    os.environ.setdefault("LC_ALL", "C.UTF-8")
    os.environ.setdefault("PYTHONIOENCODING", "utf-8")

    key_hex = os.environ.get("SIGNING_KEY_HEX")
    if not key_hex:
        print("SIGNING_KEY_HEX required", file=sys.stderr)
        sys.exit(1)
    # Load the Ed25519 private key exactly once; reuse across every record.
    private_key = Ed25519PrivateKey.from_private_bytes(bytes.fromhex(key_hex))

    while True:
        line = sys.stdin.readline()
        if not line:
            break  # EOF — exit loop, terminate daemon
        line = line.strip()
        if not line:
            continue  # skip blank keep-alive lines
        req_id = None
        try:
            req = json.loads(line)
            req_id = req.get("id")
            record = req["record"]
            canonical = canonicalize(record)
            sig = private_key.sign(canonical.encode("utf-8"))
            resp = {
                "id": req_id,
                "ok": True,
                "canonical": canonical,
                "signature_hex": sig.hex(),
            }
        except Exception as e:
            resp = {"id": req_id, "ok": False, "error": str(e)}
        sys.stdout.write(json.dumps(resp) + "\n")
        sys.stdout.flush()


def main():
    if os.environ.get("DAEMON_MODE") == "1":
        daemon_main()
        return

    # One-shot mode: byte-identical to pre-patch behavior.
    key_hex = os.environ.get("SIGNING_KEY_HEX")
    if not key_hex:
        print("SIGNING_KEY_HEX required", file=sys.stderr)
        sys.exit(1)
    private_key = Ed25519PrivateKey.from_private_bytes(bytes.fromhex(key_hex))

    record = json.loads(sys.stdin.read())
    canonical = canonicalize(record)
    sig = private_key.sign(canonical.encode("utf-8"))
    print(json.dumps({"canonical": canonical, "signature_hex": sig.hex()}))


if __name__ == "__main__":
    main()

File: test/vectors/cross-emitter-matrix/sign.rb
#!/usr/bin/env ruby
# Cross-emitter matrix — Ruby signer.
# Uses tuple-array canonical form matching ../verify.mjs, ../verify.py.
# Signing key: hex-encoded 32-byte Ed25519 seed from SIGNING_KEY_HEX env var.

# Point Ruby at the user-installed ed25519 gem
$LOAD_PATH.unshift(File.join(Dir.home, '.gem/ruby/2.6.0/gems/ed25519-1.3.0/lib'))
require 'ed25519'
require 'json'

FIELD_ORDER = %w[
  id timestamp method toolName namespace upstream
  principal durationMs success errorCode previousHash
].freeze

def assert_well_formed(str)
  str.each_char.with_index do |ch, i|
    code = ch.ord
    if code >= 0xD800 && code <= 0xDFFF
      raise "unpaired surrogate at index #{i}"
    end
  end
end

def canonicalize_value(v)
  return nil if v.nil?
  case v
  when String
    assert_well_formed(v)
    v
  when TrueClass, FalseClass
    v
  when Integer
    raise "unsafe number #{v}" if v.abs > (2**53 - 1)
    v
  when Float
    raise "unsafe number #{v}"
  when Array
    ['L', v.map { |item| canonicalize_value(item) }]
  when Hash
    keys = v.keys.sort_by { |k| k.encode('UTF-16BE').bytes }
    ['M', keys.map { |k| [k, canonicalize_value(v[k])] }]
  else
    raise "unsupported type #{v.class}"
  end
end

def canonicalize(record)
  ordered = FIELD_ORDER.map { |k| [k, canonicalize_value(record[k])] }
  %w[decisionContextDigest extensionsDigest aiInvocation parties].each do |opt|
    ordered << [opt, canonicalize_value(record[opt])] unless record[opt].nil?
  end
  JSON.generate(ordered)
end

key_hex = ENV['SIGNING_KEY_HEX']
abort 'SIGNING_KEY_HEX required' if key_hex.nil? || key_hex.empty?

# Signing key is initialized once and reused across records in daemon mode.
signing_key = Ed25519::SigningKey.new([key_hex].pack('H*'))

if ENV['DAEMON_MODE'] == '1'
  # Env pinning: fall back to UTF-8 where the driver did not already set it.
  ENV['LC_ALL'] ||= 'C.UTF-8'
  ENV['PYTHONIOENCODING'] ||= 'utf-8'
  STDIN.set_encoding('UTF-8')
  STDOUT.set_encoding('UTF-8')

  STDIN.each_line do |line|
    line = line.strip
    next if line.empty?

    # Extract id via regex on the RAW line BEFORE JSON.parse. The line may
    # contain lone surrogates (or other pathological bytes) that break
    # JSON.parse itself, in which case req['id'] would be unreachable and
    # the response would either drop entirely (breaking
    # response_count == request_count) or emit id:null (breaking id
    # correlation). The regex fallback preserves correlation even when the
    # JSON is unparseable. Note: String#match itself raises on invalid
    # UTF-8 bytes, so scrub a copy first — we still leave `line` untouched
    # for JSON.parse so it can fail authentically on the bad input.
    id_from_regex = nil
    begin
      safe_line = line.dup.force_encoding('UTF-8').scrub('?')
      id_match = safe_line.match(/"id"\s*:\s*(?:"((?:[^"\\]|\\.)*)"|(-?\d+(?:\.\d+)?))/)
      id_from_regex = id_match ? (id_match[1] || id_match[2]) : nil
    rescue
      id_from_regex = nil
    end

    begin
      req = JSON.parse(line)
      id = req['id']
      record = req['record']
      canonical = canonicalize(record)
      sig = signing_key.sign(canonical)
      STDOUT.puts JSON.generate({
        'id' => id,
        'ok' => true,
        'canonical' => canonical,
        'signature_hex' => sig.unpack('H*').first
      })
      STDOUT.flush
    rescue Exception => e
      # Paranoid catch: any Exception subclass (StandardError, ScriptError,
      # and SystemExit-family) is normalized into a single well-formed
      # response line so the driver's response_count == request_count
      # invariant holds. The fact-of-rejection (ok:false) is the scientific
      # signal — preserve it; only the message bytes are sanitized.
      raw_msg = (e.message rescue 'unknown').to_s
      sanitized_msg = begin
        raw_msg.dup.force_encoding('UTF-8').scrub('?')
      rescue
        begin
          raw_msg.dup.force_encoding('UTF-8')
                 .encode('UTF-8', invalid: :replace, undef: :replace, replace: '?')
        rescue
          'unencodable_error'
        end
      end
      begin
        STDOUT.puts JSON.generate({
          'id' => id_from_regex,
          'ok' => false,
          'error' => sanitized_msg
        })
      rescue => _nested
        # Absolute last resort: hand-built JSON line so the driver still
        # sees the rejection tag and id correlation for this request.
        STDOUT.puts %Q({"id":#{id_from_regex.to_json},"ok":false,"error":"internal_encode_failure"})
      end
      STDOUT.flush
      next
    end
  end
else
  record = JSON.parse(STDIN.read)
  canonical = canonicalize(record)
  sig = signing_key.sign(canonical)

  puts JSON.generate({ 'canonical' => canonical, 'signature_hex' => sig.unpack('H*').first })
end
